# 数値解析第３回課題

### 学籍番号：08B22182　氏名：平山聖輝

課題3：以下の行列の最大固有値 $ \lambda $ を，冪乗法で求めようとしたときの，繰り返し計算1回目での $ \lambda^{(0)} $ ，2回目の $ \lambda^{(1)} $ を求めよ．

$$
A =  \left(\begin{array}{cc}  
        4 & 2 \\ 
        1 & 5 
     \end{array}\right)
$$

ただし初期ベクトル $ x^{(0)} $ は，$ x^{(0)} = (1,0)^T $ とする．またmaxノルムとレイリー商を使った2通りの場合で求めること．

### maxノルムを用いた場合

In [9]:
import numpy as np

In [11]:
A = np.array([[4,2],[1,5]])
x0 = np.array([[1],[0]])
print("A = \n", A)
print("x0 = \n", x0)

A = 
 [[4 2]
 [1 5]]
x0 = 
 [[1]
 [0]]


$ y^{(k)} = Ax^{(k)} $ を満たすように仮の変数 $ y^{(k)} $ を設定することで $ x^{(k+1)} $ を求める．ただしノルムはmaxノルムを使う．

$$
y^{(0)} = Ax^{(0)}
$$
$$
x^{(1)} = \frac{y^{(0)}}{\|y^{(0)}\|_{\infty}}
$$
$$
\lambda^{(0)} = \|Ax^{(0)}\|_{\infty}
$$

In [41]:
y0 = A@x0
lambda0 = np.linalg.norm(y0,np.inf)
x1 = y0/np.linalg.norm(y0,np.inf)
print("x1 =\n", x1)
print("lambda0 =\n", lambda0)

x1 =
 [[1.  ]
 [0.25]]
lambda0 =
 4.0


よってmaxノルムを用いた固有ベクトルと固有値は以下

$$
\begin{equation}
 x^{(1)} =
    \left(
    \begin{array}{c}
    1.00 \\
    0.25
    \end{array}
    \right)
\end{equation}
$$
$$
 \lambda^{(0)} =4.0
 $$

$ \lambda^{(1)} $ も同様に求める．

In [44]:
y1 = A@x1
lambda1 = np.linalg.norm(y1,np.inf)
x2 = y1/lambda1
print("x2 =\n", x2)
print("lambda1 =\n", lambda1)

x2 =
 [[1. ]
 [0.5]]
lambda1 =
 4.5


$$
\begin{equation}
 x^{(2)} =
    \left(
    \begin{array}{c}
    1.00 \\
    0.5
    \end{array}
    \right)
\end{equation}
$$
$$
\lambda^{(1)} = 4.5
$$

### レイリー商を用いる場合

レイリー商では固有値 $ \lambda^{(k)} $ を

$$
\lambda^{(k)} = \|Ax^{(k)}\|
$$

ではなく，レイリー商

$$
R_A(x) = \frac{(Ax^{(k)},x)}{\|x\|^2_2} = \frac{x^TAx}{x^Tx}
$$

を用いて計算すればよい．具体的には以下の手順で求める．

$$
y^{(0)} = Ax^{(0)}
$$
$$
X^{(1)} = \frac{y^{(0)}}{\|y^{(0)}\|_2}
$$
$$
\lambda'^{(0)} = \frac{{x^{(0)}}^TAx^{(0)}}{{x^{(0)}}^Tx^{(0)}}
$$

In [55]:
X1 = y0/np.linalg.norm(y0)
Lambda0 = np.dot(x0.T,A@x0)/np.dot(x0.T,x0)
print("X1 =\n" + str(X1))
print("Lambda0 =\n" + str(Lambda0))

X1 =
[[0.9701425 ]
 [0.24253563]]
Lambda0 =
[[4.]]


よってレイリー商を用いた場合の固有値と固有値は以下．

$$
\begin{equation}
 X^{(1)} =
    \left(
    \begin{array}{c}
    0.97014250 \\
    0.24253563
    \end{array}
    \right)
\end{equation}
$$
$$
\lambda'^{(0)} = 4.0
$$

$ \lambda'^{(1)} $ も同様に求める．

In [64]:
Y1 = A@X1
X2 = Y1/np.linalg.norm(Y1)
Lambda1 = np.dot(X1.T, A@X1)/np.dot(X1.T, X1)
print("X2 =\n", X2)
print("Lambda1 =\n", Lambda1)

X2 =
 [[0.89442719]
 [0.4472136 ]]
Lambda1 =
 [[4.76470588]]


$$
X^{(2)} =
    \left(
    \begin{array}{c}
    0.89442719 \\
    0.44721360
    \end{array}
    \right)
$$
$$
\lambda'^{(1)} = 4.76470588
$$